flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,214 @@
+                                    }
+                                });
+        DataStream<KnnModelData> distributedModelData = prepareModelData(inputData);
+        DataStream<KnnModelData> modelData = mergeModelData(distributedModelData);
+        KnnModel model = new KnnModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return this.params;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static Knn load(StreamExecutionEnvironment env, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Prepares distributed knn model data. Constructs the sample matrix and computes norm of
+     * vectors.
+     *
+     * @param inputData Input vector data with label.
+     * @return Distributed knn model.
+     */
+    private static DataStream<KnnModelData> prepareModelData(
+            DataStream<Tuple2<DenseVector, Double>> inputData) {
+        return DataStreamUtils.mapPartition(
+                inputData,
+                new RichMapPartitionFunction<Tuple2<DenseVector, Double>, KnnModelData>() {
+                    @Override
+                    public void mapPartition(
+                            Iterable<Tuple2<DenseVector, Double>> values,
+                            Collector<KnnModelData> out) {
+                        List<Tuple2<DenseVector, Double>> dataPoints = new ArrayList<>(0);
+                        for (Tuple2<DenseVector, Double> tuple2 : values) {
+                            dataPoints.add(tuple2);
+                        }
+                        int featureDim = dataPoints.get(0).f0.size();
+                        DenseMatrix packedFeatures = new DenseMatrix(featureDim, dataPoints.size());
+                        DenseVector labels = new DenseVector(dataPoints.size());
+                        for (int i = 0; i < dataPoints.size(); ++i) {
+                            Tuple2<DenseVector, Double> tuple2 = dataPoints.get(i);
+                            labels.values[i] = tuple2.f1;
+                            System.arraycopy(
+                                    tuple2.f0.values,
+                                    0,
+                                    packedFeatures.values,
+                                    i * featureDim,
+                                    featureDim);
+                        }
+                        DenseVector featureNorms = computeNorm(packedFeatures);
+                        if (dataPoints.size() > 0) {
+                            out.collect(new KnnModelData(packedFeatures, featureNorms, labels));
+                        }
+                    }
+                },
+                TypeInformation.of(KnnModelData.class));
```

Comment:
It seems that we can simplify the code here by calling `mapPartition(...)` without specifying this TypeInformation. Could you double check this?

Same for `mergeModelData(...)`.

Reply:
OK